argo/common/run_completion.go

```diff
@@ -20,18 +24,57 @@ var RunCompletionStatuses = struct {
 }
 
 type RunCompletionEvent struct {
-	Status       RunCompletionStatus `json:"status"`
-	PipelineName NamespacedName      `json:"pipelineName"`
+	Status       RunCompletionStatus `json:"status" validate:"required"`
+	PipelineName NamespacedName      `json:"pipelineName" validate:"pipelineName"`
 	// Optionally render structs until https://github.com/golang/go/issues/11939 is addressed
 	RunConfigurationName  *NamespacedName `json:"runConfigurationName,omitempty"`
 	RunName               *NamespacedName `json:"runName,omitempty"`
-	RunId                 string          `json:"runId"`
+	RunId                 string          `json:"runId" validate:"required"`
 	ServingModelArtifacts []Artifact      `json:"servingModelArtifacts"`
 	Artifacts             []Artifact      `json:"artifacts,omitempty"`
-	Provider              string          `json:"provider"`
+	Provider              string          `json:"provider" validate:"required"`
 }
 
 func (sre RunCompletionEvent) String() string {
 	return fmt.Sprintf("{Status:%s PipelineName:%+v RunConfigurationName:%+v RunName:%+v RunId:%s ServingModelArtifacts:%+v Artifacts:%+v Provider:%s}",
 		sre.Status, sre.PipelineName, sre.RunConfigurationName, sre.RunName, sre.RunId, sre.ServingModelArtifacts, sre.Artifacts, sre.Provider)
 }
+
+func validateNamespacedName(nn *NamespacedName, key string) error {
+	if nn == nil {
+		return fmt.Errorf("key: %s is nil", key)
+	} else if nn.Name == "" {
+		return fmt.Errorf("key: %s, Name field is missing", key)
+	}
+
+	return nil
+}
+
+func (sre RunCompletionEvent) Validate() error {
+	validate := validator.New()
```

Comment:
This shouldnt be done here. validator.New() is thread safe (well according to the comments in the code) so should only be created once and used not created on each message

Reply:
Changed in [5bc6a9f](https://github.com/sky-uk/kfp-operator/pull/352/commits/5bc6a9f67412272114e686d4ac9d403c0dbb9ccb) which uses Mutex to control the creation of the validator and registering of custom validation rules